### PR TITLE
NBLAST score matrix calculation

### PR DESCRIPTION
**Pull request by clbarnes**
I feel that the original smat.csv can only be stretched so far. Different score matrices would probably be valuable for different brains, different tasks (lateral pairs/ lineage classification/ segmental repetition). This impl is based on my equally-untested experimental rust impl, which is based on the description of the algorithm in the paper (rather than the R impl).

### CODE REVIEW

**Comment by clbarnes** [comment]
How the dist/dot bins are specified needs work as it's hard to intuit with the infinity handling. I'm now leaning towards explicitly only asking for the interior boundaries, and adding -inf and inf to either side; could avoid doing so if they are already infs.

**Comment by schlegelp** [comment]
That's a great idea! I always wondered whether the score matrices based on fly circuit (light level, many brains) might be too forgiving when applied to EM data (within brain comparison, nm resolution, probably better transforms).

Would love to hear @jefferis' thoughts on this. For example: we have a large collection of curated neurons from the hemibrain which we could use to generate an EM-based scoring matrix.

**Comment by jefferis** [comment]
The intention was always to provide different score matrices for different circumstances, just like BLAST. Specifically you are supposed to use different BLAST substitution matrices depending on how distantly related you think your proteins may be. However also just like BLAST the default does sufficiently well that there has not been much impetus to make new matrices. Unlike BLAST there is second factor besides relatedness to consider, which is spatial scale. This is potentially quite important for the larval vs adult difference and would be a good reason to try regenerating a matrix there. In the past I've told people working on brains other than adult fly to try scaling their neurons up or down by a factor of eg 2x or 0.5x for larva or zebrafish, respectively.

In terms of EM vs LM, I don't expect much of an issue. Our registration of the flycircuit dataset was very high quality. The registration across two EM datasets or between l and r of a single dataset will not be substantially better. This is further emphasised by the fact that there was an uncorrected Z compression artefact in the flycircuit data meaning that stuff is a little closer there than in real life.

What will make a significant difference is clustering neurons in the same EM hemisphere. If you want to make an adult scoring matrix for that purpose, then the distances will likely be appreciably smaller due to cofasciculation. However even there I have some concerns that you might end up rewarding the very tight fasciculation of long tracts too much if you do not train based on very similar cell type differences (which will be in the axons and dendritic arbours).

Final point, the original matrices were only constructed with olfactory uPNs. There is no doubt that some greater diversity of cell types would be good.

**Comment by jefferis** [comment]
@dokato FYI

**Comment by jefferis** [comment]
@clbarnes in terms of the dotprops upper and lower bounds, I guess extending them to +/- Inf is no problem. In R I clamped the range when calculating them for each comparison to (0,1) using the zapsmall function. I guess there is a very small computational cost to doing this which could be avoided by range extension.

**Comment by clbarnes** [comment]
Thanks all! Indeed, I've seen some scaling of the score matrix for different sizes of animal, and as you say it would react to cell types differently. I imagine it could also be impacted by how you pre-process the neurons.

Still have yet to assemble some training data for testing this, will let you know when I have.

**Comment by schlegelp** [comment]
Just a side note but I wonder if it would be possible to implement `ScoreMatrixBuilder` in a way that we can produces scoring matrices of arbitrary dimension? For example, we might want to bring in a third metric in addition to distance and vector product.

**Comment by jefferis** [comment]
> For example, we might want to bring in a third metric in addition to distance and vector product.
This is definitely of interest but I have somewhat resisted it in the past because you quickly end up with limited data feeding in some cells (it's possible that one might need to fit a smooth surface and use that etc).

**Comment by clbarnes** [comment]
Shouldn't be too hard! It'll probably be most flexible to abstract it away from the dist/dot/alpha terminology (just take a `Callable[[Dotprops, Dotprops], List[int]]` for generating axis indices), then a subclass could reintroduce the concrete case.

It might be nice to have the same structure reflected on the `ScoreFunction` side: the Nblaster generic over `Callable[[Dotprops, Dotprops], float]` as a score function, have ND lookup tables implement `__call__` like the `ScoreFunction` does currently, and separate the lookup/ surface fit/ passthrough cases rather than handling them all in one object. A `ScoreLookup2D` would be a special case which could have `from_csv` and `from_pandas` class methods, so that its use is explicit rather than being overly flexible in the constructor which would make subclassing harder.

**Comment by clbarnes** [comment]
ND case is covered, as is an ND lookup class, and a 2D lookup class which inherits from it and provides pandas IO, plus some convenience cases. Still yet to check whether it actually works...

**Comment by clbarnes** [comment]
47 unit tests added (OK, I cheated, they're parameterised), and the score matrix produced by a synthetic example (matching the neurons already in the repo with their jittered selves) looks like the right shape.

**Comment by schlegelp** [comment]
Another one for the wish list: it would be super useful to have some metric(s) to quantify how well a given scoring function/matrix is doing given the list of matching and non-matching neurons.

nbl_matrix

Not sure if this would necessarily be part of `LookupNd` but some way to compare/evaluate scoring matrices and functions would be useful.

**Comment by schlegelp** [comment]
I also wonder if the `ScoringFunction` class should be moved from `navis.nbl.nblast_funcs` to `navis.nbl.smat` to have everything scoring-related in one place. Similarly `ScoringFunction` would need to be modified to allow N-dimensional look-up.

@clbarnes I'm happy to merge unless you want to first play around with the above.

**Comment by clbarnes** [comment]
> Another one for the wish list: it would be super useful to have some metric(s) to quantify how well a given scoring function/matrix is doing given the list of matching and non-matching neurons.

Would this going to be a measure of the score function or might it fit better on the NBlaster wrapping it? As I imagine it's not too clear how useful it'll be until you've done the summing and normalisation. I guess you might want to just get the raw distdot scores out for some given pairs of neurons so you could then do some separability tests of your own or show them on a violin plot or something (but the noise would be pretty large).

> I also wonder if the ScoringFunction class should be moved from navis.nbl.nblast_funcs to navis.nbl.smat to have everything scoring-related in one place. Similarly ScoringFunction would need to be modified to allow N-dimensional look-up.

Yes, I haven't got round to looking at ScoreFunction yet. I wonder if it might be better to deprecate it entirely (with `(Pending)DeprecationWarning`) - if people want to use the passthrough case, they should just use `operator.mul`, and if they want a lookup table, there's `Lookup2D`, which has the same pandas interop. I could make `ScoringFunction` a wrapper around those two options for compatibility. I opted to just implement to/from dataframe on Lookup2D (rather than add file reading etc.) because pandas' IO is very flexible and it didn't seem worth re-implementing a subset of that when users could just `pd.read/to_whatever` themselves.

**Comment by schlegelp** [comment]
> Would this be a measure of the score function or might it fit better on the NBlaster wrapping it?

To my mind this would be most useful if it was flexible. If it was a separate function that takes NBLAST scores + a list of matching pairs and produces a couple interesting metrics (e.g. min/mean/max within and across type scores), it could be plugged in at the end of `LookupNdBuilder` as well as after a proper normalized NBLAST. 

> Yes, I haven't got round to looking at ScoreFunction yet. I wonder if it might be better to deprecate it entirely.

I think that's fair: I'd be surprised if anybody other than us used it in it's current form. Most users won't interact with it anyway. As things are implemented at the moment, there isn't even a way of using e.g. `navis.nblast` with a custom scoring function. On that end, I had imagined that the nblast functions would get a new parameter `scoring_func` that looks something like this:

```
    scoring_func :  str | pandas.DataFrame | Lookup2d | callable, optional
                    A scoring function or matrix:
                      - strings will be passed to `pandas.read_csv` to produce a scoring matrix
                      - DataFrames are expected
                      - an instance of `navis.nbl.smats.NDLookup`
                      - a callable that accepts two (N, ) arrays of distances of
                        nearest-neighbor vector pairs and their dot products and
                        returns a single (N, ) array of the final scores
                      - `None` will use the original scoring matrices from the 
                        NBLAST paper
```

Also just to make a note: once this is stable enough, it be nice to add a tutorial on "making your own NBLAST scoring function" to the docs.

File: navis/nbl/nblast_funcs.py [hunk 1]
```diff
@@ -124,15 +147,24 @@ def __init__(self, use_alpha=False, normalized=True, smat='auto', progress=True)
...
-                                   index_col=0)
+                smat = smat_path / 'smat_fcwb.csv'
+
+        if isinstance(smat, (str, os.PathLike)):
+            smat = pd.read_csv(smat, index_col=0)
+
+        if isinstance(smat, pd.DataFrame):
+            smat = Lookup2d.from_dataframe(smat)
+
+        if not callable(smat):
```
Comment (schlegelp) [inline navis/nbl/nblast_funcs.py:164]: I'm a big fan of very verbose error messages (for dummies like myself): in case somebody passed a `callable` as `smat`, it might be a good idea to test if it's valid: i.e. it accepts two (N, ) numpy arrays and returns a single (N, 3) array.

Comment (clbarnes) [reply navis/nbl/nblast_funcs.py:164]: Completely agree on failing quickly and explicitly, and this would work fine for the moment. We'd need to consider how to keep that strictness in a possible arbitrary N-dimensional score function future, where we don't necessarily know how many arguments it should take (although we could use introspection), or what the domain of the function is (if it's only defined within certain values, we don't want to refuse to use it just because we don't know what those values are).

But a fair bit would have to change elsewhere in the code for us to even get to that future, so we can cross that bridge when we come to it.

Comment (schlegelp) [reply navis/nbl/nblast_funcs.py]: Sweet! In theory, the `NBlaster` class (and any variants of it like `SynBlaster`) should "know" what they expect from the scoring function. Maybe it would be most convenient to make it a class method like `validate_score_func(self)`?

Comment (schlegelp) [reply navis/nbl/nblast_funcs.py]: Also: I feel like I'm making you jump through too many hoops here. I'm perfectly happy with what's there right now and would not mind just merging if your commits already cover your use case.

**Comment by clbarnes** [comment]
Last commit is a bit speculative - re-adds helper methods on the `LookupDistDotBuilder` for determining the boundaries. This is trivial for the dots, but less so for the dists: settled on asking for another argument which sets the scaling, but I suspect it won't do very well in practice.

The right way to do this may be to wait until you have all the distdots and then partition it like a balanced tree.

You could just keep a (massively downsampled) actual spatial tree as a score function, of course (because what this algorithm really needs is more nearest neighbour queries...).

**Comment by schlegelp** [comment]
I guess that this would be a case where one might run a couple iterations to find good bounds - for which you need a way to evaluate the results :)

**Comment by schlegelp** [comment]
Just a thought sparked by #38: I was wondering if it would make sense to nudge users to follow a convention by which matrices are always made for data in microns. Minimally this could be something in the docstring, or perhaps a warning if the dotprops are not in microns.

**Comment by schlegelp** [comment]
@clbarnes not sure if you have met @bdpedigo before:

He is a grad student in Joshua Vogelstein's lab at John Hopkins. They are doing very cool work matching neurons across data sets (plus lots of other cool stuff that I'm just not aware of). I wonder if Ben _et al._ might be have thoughts/ideas on how to improve NBLAST including but not limited to scoring functions.

**Comment by clbarnes** [comment]
Yes, Ben's spent some time in the lab with us and is working with Michael Winding on the whole brain project, on the graph theory end of things. Unfortunately the last visit was cut short by the impending lockdown 1 last year.

**Comment by schlegelp** [comment]
🤦 Ah right - I actually remember seeing him at the LMB.

**Comment by bdpedigo** [comment]
> I wonder if Ben et al. might be have thoughts/ideas on how to improve NBLAST including but not limited to scoring functions. I'd be particularly interested whether you have any thoughts on how to assess "quality" of a scoring matrix: i.e. given a set of known matches and non-matches, how good is the NBLAST + scoring function at finding the correct matches and separating non-matches.

This is actually sort of why I wanted the `smat` kwarg, I have been playing around with different transformations of the raw NBLAST scores. I found one that I liked at least at a glance in terms of clustering - not sure how it compares to what you'd usually run, though. If you have a set of pairs that you like, I think we'd definitely have some ideas for assessing the different scoring functions (off the top of my head I'd say we'd usually use something like mean reciprocal rank or precision/recall @ k). Happy to chat more. Would you have one correct match for each neuron, or possibly multiple?

> Unfortunately the last visit was cut short by the impending lockdown 1 last year. 

Crazy that it's been a whole year!

**Comment by clbarnes** [comment]
It may depend on the task the classifier is being used for. The use cases we're invested in are

- finding a homologous neuron from one brain in another brain
- finding the partner in a lateral pair (by transforming the neuron)
- finding the equivalent neuron in a different segment (also by transforming)
- clustering neurons by lineage (probably needs some other metrics too)

I suspect the lineage case may want slightly different tuning to the other 3. In any case we'd probably want to rank potential matches as you say, with a p value for each.

**Comment by bdpedigo** [comment]
> It may depend on the task the classifier is being used for. The use cases we're invested in are
> 
> * finding a homologous neuron from one brain in another brain
> * finding the partner in a lateral pair (by transforming the neuron)
> * finding the equivalent neuron in a different segment (also by transforming)
> * clustering neurons by lineage (probably needs some other metrics too)
> 
> I suspect the lineage case may want slightly different tuning to the other 3. In any case we'd probably want to rank potential matches as you say, with a p value for each.

That all makes sense. If you all have any 'training data' so to speak for the above tasks and/or just want to chat more about it just let me know!

**Comment by schlegelp** [comment]
Just to leave a paper trail for this: @dokato has brought up [Dunn Index](https://en.wikipedia.org/wiki/Dunn_index) or [Rand Index](https://en.wikipedia.org/wiki/Rand_index) to assess clustering.

File: navis/nbl/smat.py [hunk 2]
```diff
@@ -0,0 +1,447 @@
...
+        ----------
+        strict : bool, optional
+            If falsey (default), ignores parentheses and brackets,
+            effectively normalising to
+            ``{IMPLICIT_INTERVAL[0]}lower,upper{IMPLICIT_INTERVAL[1]}``
+            as an implementation detail.
+            Otherwise, raises a ValueError if parens/brackets
+            do not match the implementation.
+        """
+        boundaries = [
```
Comment (clbarnes) [inline navis/nbl/smat.py:439]: Do this in one iteration

File: navis/nbl/smat.py [hunk 3]
```diff
@@ -0,0 +1,447 @@
...
+    if strict and item[0] + item[-1] != IMPLICIT_INTERVAL:
+        raise ValueError(
+            f"Enclosing characters {explicit_interval} "
+            f"do not match implicit interval specifiers {IMPLICIT_INTERVAL}"
+        )
+    return tuple(float(i) for i in item[1:-1].split(","))
+
+
+def parse_boundaries(items, strict=False):
+    last = None
```
Comment (clbarnes) [inline navis/nbl/smat.py:392]: `last` is unnecessary, just use `upper`

File: navis/nbl/nblast_funcs.py [hunk 4]
```diff
@@ -124,18 +153,44 @@ def __init__(self, use_alpha=False, normalized=True, smat='auto', progress=True)
...
 
-        self.self_hits = []
-        self.dotprops = []
+        if isinstance(smat, pd.DataFrame):
+            smat = Lookup2d.from_dataframe(smat)
+
+        if not callable(smat):
+            raise ValueError("smat should be a callable, a path, a pandas.DataFrame, or 'auto'")
+
+        if not isinstance(smat(0.5, 0.5), float):
```
Comment (clbarnes) [inline navis/nbl/nblast_funcs.py:179]: Refactor into private function

**Review by clbarnes — COMMENTED** [review]
Some minor readability todos (as well as rebases)

File: navis/nbl/nblast_funcs.py [hunk 5]
```diff
@@ -40,13 +44,20 @@ class ScoringFunction:
```
Comment (clbarnes) [inline navis/nbl/nblast_funcs.py:44]: Note deprecation in docstring.

**Comment by clbarnes** [comment]
On rebase, `smat` will be in the `nblast_*` free functions so the common smat-interpreting bits should be factored out.

**Comment by clbarnes** [comment]
The builder tests are hitting a deadlock when run together, fine when run individually.

File: navis/nbl/smat.py (hunk 3 above)
Comment (clbarnes) [reply navis/nbl/smat.py]: Still have to do a slightly long way round for type checking and 0-length reasons.

File: navis/nbl/nblast_funcs.py (hunk 1 above)
Comment (clbarnes) [reply navis/nbl/nblast_funcs.py]: There is now a functional function-checker, which is factored into its own function which takes as an argument the number of arguments the checked function should take. Phew, that's a mouthful. Anyway, it would be easy enough to check that at the class level in future.

**Comment by clbarnes** [comment]
Rebased.

The last couple of commits refactor more of the score matrix-related things into the `smat` module; in particular, there is now a `smat_fcwb` function for getting the vendored score matrices, and `(parse|check)_score_fn` functions for backwards compatibility. @schlegelp , any thoughts on which `smat.py` members should be exported in `__all__`?

**Comment by schlegelp** [comment]
> @schlegelp , any thoughts on which smat.py members should be exported in __all__?

Given that they are very specialised functions/classes, I would have probably exposed `Lookup2d`, `LookupNd`, `smat_fcwb smat_fcwb` and `parse_score_fn` at module level (i.e. `navis.nbl.LookupNd`, not `navis.LookupNd`) and left the builders in `navis.nbl.smat` for "power users".

**Comment by schlegelp** [comment]
Hi @clbarnes. I'm doing some work on the whole nblast module and I was wondering if you wanted to merge this PR to avoid having to making a huge rebase further down the line?

**Comment by clbarnes** [comment]
Sorry to have let this go stale - I think the rebase will be significant so I wouldn't wait on this before 1.0.

**Comment by clbarnes** [comment]
Closing in favour of #69
